### PR TITLE
feat: v0.13.0 Git Archaeology — retroactive decision extraction

### PR DESCRIPTION
**Pull request by teslamint**
## Summary

- **`ec archaeologize` command** — streams `git log --patch` through the existing extraction pipeline to produce `source:archaeology` decision candidates from commit history. Supports `--since`, `--until`, `--limit` (default 100), `--pr-bodies` (GitHub API), `--dry-run` (cost estimate), `--batch-size`. Implicit resume via dedup.
- **Schema v16 migration** — `archaeology_processed` table for commit-level dedup + `decision_candidates.source_type` CHECK expansion to accept `'archaeology'`. FTS triggers properly rebuilt.
- **`run_extraction` bundles seam** — optional `bundles` parameter and `session_id: str | None` allow external callers to inject pre-built `SignalBundle` objects while sharing the full extraction pipeline. Single entry point, no pipeline fork.
- **`[decisions.archaeology]` config section** — `enabled`, `batch_size`, `pr_body_fetch` settings.
- Addresses the **91% file-link gap** (122 decisions, only 11 with file links) and eliminates the **cold-start adoption barrier**.

## Key design decisions

- `run_extraction` receives a minimal seam (not a parallel extraction path) to avoid pipeline drift — per LESSONS.md: "공유 helper로 policy divergence 방지"
- Commit message (`%B` full body) prepended to bundle text for LLM context
- `archaeology_processed` table tracks all processed commits (including zero-candidate ones) for idempotent re-runs
- `--pr-bodies` requires explicit flag; token passed via `GH_TOKEN` env to `gh api`
- `--limit N + --reverse` selects most-recent-N (documented as known limitation)

## Test plan

- [ ] Schema v16 migration tests (6): table creation, CHECK expansion, data preservation, FTS triggers, rejected values
- [ ] Extraction seam tests (6): type registration, bundle construction, injected bundles, NULL session_id persistence
- [ ] Core archaeology tests (20+): helpers, dry-run, processing, dedup resume, interrupt recovery, commit message in bundle
- [ ] Integration tests (4): full pipeline, re-run dedup, dry-run no-writes, source_type assertion
- [ ] CLI tests (2): help output, dry-run
- [ ] Full suite: 1946 passed, 1 skipped (pre-existing), 0 failed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_012qKsBWxCBxsK1oLCKUYA1v

### CODE REVIEW

File: src/entirecontext/core/archaeology.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        bundle = _build_signal_bundle(sha, message, patch_text, pr_body)
+        try:
+            outcome = run_extraction(
+                conn,
+                session_id=None,
+                repo_path=repo_path,
+                bundles=[bundle],
+                extraction_weights=extraction_weights,
+            )
+            _mark_processed(conn, sha, outcome.candidates_inserted)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in b5a69f5. `_process_batch` now only calls `_mark_processed` when `outcome.parsed_ok or outcome.candidates_inserted > 0`. Failed extractions get a retry warning instead of being permanently marked done. Added regression test `test_failed_extraction_not_marked_processed`.

File: src/entirecontext/core/archaeology.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    # on \x00 with maxsplit=2 to get (sha, message, patch). %B (not %s) is
+    # used so the full commit body reaches the bundle, not just the subject.
+    cmd = ["git", "log", "--patch", "--reverse", "--format=%x1e%H%x00%B%x00"]
+    if since and _looks_like_date(since):
+        cmd.append(f"--since={since}")
+    elif since:
+        rev_range = f"{since}..{until}" if until and not _looks_like_date(until) else f"{since}..HEAD"
+        cmd.append(rev_range)
+    if until and _looks_like_date(until):
+        cmd.append(f"--until={until}")
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in b5a69f5. `_stream_commits` rewritten to classify `since`/`until` independently as ref-or-date, covering all 4 combinations. `--until v1.0.0` without `--since` now correctly passes the ref to git log. Added tests for all ref/date permutations.

File: src/entirecontext/core/archaeology.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+    result = subprocess.run(
+        cmd,
+        cwd=repo_path,
+        capture_output=True,
+        text=True,
+        errors="surrogateescape",
+    )
+    if result.returncode != 0:
+        return
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:154]: Fixed in b5a69f5. `_stream_commits` now takes an optional `warnings` list and reports git log failures (exit code + stderr) instead of silently returning empty. Added test `test_git_log_failure_reported`.

File: src/entirecontext/core/archaeology.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    cmd = ["git", "log", "--patch", "--reverse", "--format=%x1e%H%x00%B%x00"]
+    if since and _looks_like_date(since):
+        cmd.append(f"--since={since}")
+    elif since:
+        rev_range = f"{since}..{until}" if until and not _looks_like_date(until) else f"{since}..HEAD"
+        cmd.append(rev_range)
+    if until and _looks_like_date(until):
+        cmd.append(f"--until={until}")
+    if limit:
+        cmd.extend(["-n", str(limit)])
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in b5a69f5. Changed `if limit:` to `if limit is not None and limit > 0:`. CLI now rejects `--limit <= 0` with an error. Added tests for the `limit=0` edge case.

File: src/entirecontext/cli/archaeology_cmds.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        console.print(
+            "[yellow]Archaeology is disabled in config. Set [decisions.archaeology] enabled = true.[/yellow]"
+        )
+        raise typer.Exit(1)
+
+    effective_batch_size = batch_size or arch_config.get("batch_size", 10)
+
+    conn = get_db(repo_path)
+    try:
+        check_and_migrate(conn)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/cli/archaeology_cmds.py]: Fixed in b5a69f5. `check_and_migrate(conn)` moved inside `if not dry_run:` block. `archaeologize()` wraps the already-processed count in `try/except sqlite3.OperationalError` for dry-run against a never-migrated DB. Added test `test_dry_run_no_migration`.

File: src/entirecontext/core/archaeology.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        capture_output=True,
+        text=True,
+    )
+    if remote_url.returncode != 0:
+        return None
+    url = remote_url.stdout.strip()
+    match = re.search(r"[:/]([^/]+/[^/.]+?)(?:\.git)?$", url)
+    if not match:
+        return None
+    owner_repo = match.group(1)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in b5a69f5. Added `if "github.com" not in url: return None` check before parsing owner/repo. Also relaxed the repo regex from `[^/.]+` to `[^/]+?` to support dotted repo names. Added tests for non-GitHub origins and dotted repo names.

File: src/entirecontext/core/archaeology.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    text_blocks = []
+    if message:
+        text_blocks.append(message)
+    if pr_body:
+        text_blocks.append(pr_body)
+    if patch_text:
+        text_blocks.append(patch_text)
+    return SignalBundle(
+        source_type="archaeology",
+        source_id=commit_sha,
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:72]: Deferred to v0.13.1. Linking archaeology candidates to `decision_commits` on confirmation requires changes to `confirm_candidate` in `decision_candidates.py` — it currently looks up checkpoint git hashes, not `source_id`. This is an architectural change beyond the current PR scope. Tracked as a follow-up: when confirming an archaeology candidate, auto-insert `(promoted_decision_id, source_id)` into `decision_commits`.

File: src/entirecontext/cli/archaeology_cmds.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            pr_bodies=pr_bodies,
+            dry_run=dry_run,
+            batch_size=effective_batch_size,
+            progress_callback=_progress,
+        )
+    finally:
+        conn.close()
+
+    if dry_run:
+        return
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/cli/archaeology_cmds.py:92]: Fixed in 7b04186. Warnings block moved before the dry-run early return — `result.warnings` (including git log failures) are now printed regardless of `--dry-run`.

File: src/entirecontext/core/archaeology.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+            capture_output=True,
+            text=True,
+            timeout=10,
+            env=env,
+        )
+        if result.returncode == 0 and result.stdout.strip():
+            return result.stdout.strip()
+    except (FileNotFoundError, subprocess.TimeoutExpired):
+        pass
+    return None
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 7b04186. `_process_batch` now appends a warning when `_fetch_pr_body` returns `None` with `--pr-bodies` enabled: `commit {sha}: PR body fetch returned empty (rate limit, permissions, or no associated PR)`. Users see the count and first 5 warnings in CLI output.

File: src/entirecontext/core/archaeology.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    remote_url = subprocess.run(
+        ["git", "remote", "get-url", "origin"],
+        cwd=repo_path,
+        capture_output=True,
+        text=True,
+    )
+    if remote_url.returncode != 0:
+        return None
+    url = remote_url.stdout.strip()
+    if "github.com" not in url:
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 6a4ec3c. Replaced `"github.com" in url` substring check with `_is_github_remote(url)` that parses the URL and requires the host to be exactly `github.com`. Rejects `notgithub.com`, `github.com.evil`, `api.github.com`. Added 7 test cases.

File: src/entirecontext/core/archaeology.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    warnings: list[str] = field(default_factory=list)
+
+
+_DIFF_HEADER_RE = re.compile(r'^diff --git "?a/.+ "?b/(.+?)"?$', re.MULTILINE)
+
+
+def _extract_files_from_patch(patch_text: str) -> list[str]:
+    if not patch_text:
+        return []
+    return list(dict.fromkeys(_DIFF_HEADER_RE.findall(patch_text)))
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 6a4ec3c. Added `_decode_git_quoted_path` that decodes Git's octal escapes (`\\355\\225\\234` → `한`) via latin-1→UTF-8 round-trip. `_extract_files_from_patch` now returns decoded paths so file links match during retrieval. Added 3 test cases.

File: src/entirecontext/core/archaeology.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+    if limit is not None and limit > 0:
+        cmd.extend(["-n", str(limit)])
+
+    result = subprocess.run(
+        cmd,
+        cwd=repo_path,
+        capture_output=True,
+        text=True,
+        errors="surrogateescape",
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:148]: Deferred to v0.13.1. Switching from `subprocess.run` to streaming `Popen` with incremental parsing is a significant refactor that changes the control flow of `_stream_commits`, `archaeologize`, and the interrupt handler. Current `--limit 100` default keeps memory bounded for typical use. Tracked as a follow-up for large-repo optimization.

File: src/entirecontext/core/archaeology.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+    token: str | None,
+    extraction_weights: ExtractionWeights | None,
+    progress_callback: Callable[[str], None] | None,
+) -> None:
+    for sha, message, patch_text in batch:
+        pr_body = None
+        if pr_bodies and token:
+            pr_body = _fetch_pr_body(sha, repo_path, token)
+            if pr_body is None:
+                result.warnings.append(f"commit {sha[:12]}: PR body fetch returned empty (rate limit, permissions, or no associated PR)")
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 9842dcf. `_process_batch` now tracks consecutive PR body fetch failures. After 3 consecutive `None` returns, PR body fetching is auto-disabled for remaining commits with a warning. Successful fetches reset the counter.

File: src/entirecontext/core/archaeology.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+
+def _mark_processed(conn: sqlite3.Connection, commit_sha: str, candidate_count: int) -> None:
+    conn.execute(
+        "INSERT OR IGNORE INTO archaeology_processed (commit_sha, candidate_count) VALUES (?, ?)",
+        (commit_sha, candidate_count),
+    )
+
+
+def _looks_like_date(ref: str) -> bool:
+    return bool(re.match(r"^\d{4}-\d{2}-\d{2}", ref))
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 9842dcf. Changed `_looks_like_date` regex from `^\d{4}-\d{2}-\d{2}` (prefix match) to `^\d{4}-\d{2}-\d{2}$` (full match). Tags like `2025-01-01-release` are now treated as refs, not dates.

File: src/entirecontext/core/archaeology.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+        record = record.strip()
+        if not record:
+            continue
+        parts = record.split("\x00", maxsplit=2)
+        if len(parts) < 2:
+            continue
+        sha = parts[0].strip()
+        message = parts[1].strip()
+        patch_text = parts[2] if len(parts) > 2 else ""
+        if len(sha) == 40:
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 9842dcf. SHA length check changed from `len(sha) == 40` to `len(sha) in (40, 64)` to support Git SHA-256 repositories.

File: src/entirecontext/core/archaeology.py [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+@dataclass
+class ArchaeologyResult:
+    commits_scanned: int = 0
+    commits_processed: int = 0
+    commits_skipped: int = 0
+    candidates_generated: int = 0
+    warnings: list[str] = field(default_factory=list)
+
+
+_DIFF_HEADER_RE = re.compile(r'^diff --git "?a/.+ "?b/(.+?)"?$', re.MULTILINE)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:28]: Deferred to v0.13.1. Paths containing ` b/` are an extreme edge case — Git itself has trouble with these in some contexts. The current regex handles all normal file paths correctly. A proper fix requires switching to `--raw` diff format or `-z` null-delimited output, which is a larger refactor.

File: src/entirecontext/core/archaeology.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+) -> ArchaeologyResult:
+    result = ArchaeologyResult()
+    token = None
+    if pr_bodies:
+        token = _get_github_token()
+        if not token:
+            result.warnings.append(
+                "No GitHub token found. Set EC_GITHUB_TOKEN or install gh CLI. Proceeding without PR bodies."
+            )
+            pr_bodies = False
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:250]: Deferred to v0.13.1. Current behavior: if no token is found, `--pr-bodies` downgrades to patch-only with a clear warning. Aborting entirely would be a UX regression — users may prefer partial extraction over none. The fix for retry (clearing processed status for tokenless runs) requires tracking `processed_with_pr_bodies` metadata, which is beyond current scope.

File: src/entirecontext/core/archaeology.py [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+    commits and then reverses their order for output — it does not walk
+    from the oldest commit forward. Re-running with the same `limit` can
+    never advance into older history; callers must increase `limit` (or
+    narrow `since`/`until`) to reach commits beyond the most recent window.
+    """
+    # %x1e (record separator) before each commit disambiguates patch content
+    # from the next commit's header. Split on \x1e, then split each record
+    # on \x00 with maxsplit=2 to get (sha, message, patch). %B (not %s) is
+    # used so the full commit body reaches the bundle, not just the subject.
+    cmd = ["git", "log", "--patch", "--reverse", "--format=%x1e%H%x00%B%x00"]
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Deferred to v0.13.1. Merge commits require `-m --first-parent` to produce diffs, plus a decision on how to attribute files. For now, the PR body (if fetched via `--pr-bodies`) provides the rationale even without the diff. Adding `--first-parent` would change scan behavior for all commits, not just merges.

File: src/entirecontext/cli/archaeology_cmds.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+        )
+        raise typer.Exit(1)
+
+    if limit <= 0:
+        console.print("[red]--limit must be a positive integer.[/red]")
+        raise typer.Exit(1)
+
+    effective_batch_size = batch_size or arch_config.get("batch_size", 10)
+
+    conn = get_db(repo_path)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/cli/archaeology_cmds.py]: Fixed in b98d7d7. Dry-run now uses `get_memory_db()` instead of `get_db(repo_path)`, so `--dry-run` never creates `.entirecontext/db/` or `local.db` on disk. Existing test `test_dry_run_does_not_create_db` already verifies this behavior.

File: src/entirecontext/cli/archaeology_cmds.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    if limit <= 0:
+        console.print("[red]--limit must be a positive integer.[/red]")
+        raise typer.Exit(1)
+
+    effective_batch_size = batch_size or arch_config.get("batch_size", 10)
+
+    if dry_run:
+        from ..db import get_memory_db
+
+        conn = get_memory_db()
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/cli/archaeology_cmds.py]: Fixed in 24b3f13. Dry-run now opens existing `local.db` read-only (`?mode=ro` URI) when present, so processed counts are accurate. Falls back to in-memory DB only when no DB exists.

File: src/entirecontext/core/archaeology.py [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+    extraction_weights: ExtractionWeights | None,
+    progress_callback: Callable[[str], None] | None,
+) -> None:
+    consecutive_pr_failures = 0
+    for sha, message, patch_text in batch:
+        pr_body = None
+        if pr_bodies and token and consecutive_pr_failures < _PR_BODY_FAIL_THRESHOLD:
+            pr_body = _fetch_pr_body(sha, repo_path, token)
+            if pr_body is None:
+                consecutive_pr_failures += 1
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 24b3f13. `_fetch_pr_body` now returns `""` (empty string) when the API succeeds but no PR/body exists, vs `None` for hard failures. Only `None` counts toward the consecutive failure threshold, so commits without PRs don't exhaust the retry budget.

File: src/entirecontext/core/archaeology.py [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+                if consecutive_pr_failures >= _PR_BODY_FAIL_THRESHOLD:
+                    result.warnings.append(
+                        f"PR body fetch failed {_PR_BODY_FAIL_THRESHOLD} times consecutively — disabling for remaining commits"
+                    )
+            else:
+                consecutive_pr_failures = 0
+
+        bundle = _build_signal_bundle(sha, message, patch_text, pr_body)
+        try:
+            outcome = run_extraction(
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py:367]: Fixed in 24b3f13. `archaeologize()` now accepts `min_confidence` parameter (default 0.35). CLI loads it from `decisions.candidate_min_confidence` or `decisions.archaeology.min_confidence` in config, matching the existing session extraction behavior.

File: src/entirecontext/core/archaeology.py [hunk 23]
```diff
@@ -0,0 +1,376 @@
...
+    commits and then reverses their order for output — it does not walk
+    from the oldest commit forward. Re-running with the same `limit` can
+    never advance into older history; callers must increase `limit` (or
+    narrow `since`/`until`) to reach commits beyond the most recent window.
+    """
+    # %x1e (record separator) before each commit disambiguates patch content
+    # from the next commit's header. Split on \x1e, then split each record
+    # on \x00 with maxsplit=2 to get (sha, message, patch). %B (not %s) is
+    # used so the full commit body reaches the bundle, not just the subject.
+    cmd = ["git", "log", "--patch", "--reverse", "--format=%x1e%H%x00%B%x00"]
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 24b3f13. `git log` command now includes `--no-color --src-prefix=a/ --dst-prefix=b/` to force stable diff header format regardless of user's `color.ui` or `diff.noprefix` git config.

File: src/entirecontext/core/archaeology.py [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+    token: str | None,
+    min_confidence: float,
+    extraction_weights: ExtractionWeights | None,
+    progress_callback: Callable[[str], None] | None,
+) -> None:
+    consecutive_pr_failures = 0
+    for sha, message, patch_text in batch:
+        pr_body = None
+        if pr_bodies and token and consecutive_pr_failures < _PR_BODY_FAIL_THRESHOLD:
+            raw_pr = _fetch_pr_body(sha, repo_path, token)
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/archaeology.py]: Fixed in 8624866. `consecutive_pr_failures` is now initialized in `archaeologize()` and passed to/returned from `_process_batch()` as a parameter. The counter persists across batches, so once the threshold is hit, PR fetching stays disabled for the rest of the run.

File: src/entirecontext/core/config.py [hunk 25]
```diff
@@ -132,6 +132,11 @@
             "contradicted_penalty": 0.15,
         },
         "auto_embed": True,
+        "archaeology": {
```
[the remark this answers is not in the document]
Comment (teslamint) [reply src/entirecontext/core/config.py:135]: Fixed in 8624866. Added `decisions.archaeology` to the `Sections:` line in CLAUDE.md.
